### ISSUE TITLE
'numpy.datetime64' object has no attribute 'year' writing from grib2 source

### ISSUE THREAD

**Issue by rsignell-usgs** [fsspec/kerchunk#130]
I have a consolidated JSON with a "best time series" dataset that reads from a collection of HRRR grib2 files on S3.   

The data access and visualization of data in the notebook looks fine, but when I attempt to save a subset (to either zarr or netcdf) I get an error from xarray:
```
AttributeError: 'numpy.datetime64' object has no attribute 'year'
```
I'm raising this issue here because I run a nearly identical workflow with NetCDF files, where we also have datetime64 objects for time in xarray, and don't get this problem.  Hence I'm thinking it must have something to do with the reading from grib2 files?

This notebook should be reproducible (and doesn't require AWS credentials):
https://nbviewer.org/gist/rsignell-usgs/7d0a76a5f9fbf5781e9e97b66bd72882

**Comment by martindurant** [comment]
The `time` column should definitely have been inlined in the references
```
Bytes | 1.38 kiB | 8 B
```
Maybe this is the problematic one, not valid_time, the coordinate, so you can remove it?

**Comment by martindurant** [comment]
Can you also try multiplying by 1.0?

**Comment by martindurant** [comment]
(or subtracting timedelta(0) )

**Comment by rsignell-usgs** [comment]
@martindurant , I'm not sure I understand.  Multiply what by 1.0?  
The datetime64 values are understood perfectly well in xarray, just not when saved to disk.
I don't understand why.  Do you?

**Comment by martindurant** [comment]
I don't know why exactly, but I suspect that if you perform an arithmetic operation (like adding a timedelta zero), then any leftover encoding attributes might be usefully forgotten.

**Comment by rsignell-usgs** [comment]
@martindurant I'm still not sure what I'm supposed to multiply by 1.0 and at what point I'm supposed to do it. 
Did you try running the notebook?   https://nbviewer.org/gist/5d10137059a1793f2606e39a5baf1ba2

**Comment by martindurant** [comment]
"forbidden" reading the JSON file. Do I need to run this in qhub?

**Comment by martindurant** [comment]
```
ds2 = ds.assign(time=ds.time + pd.Timedelta(0))
ds2.isel(time=slice(-3,-1)).to_zarr('foo.zarr', 'w')
```
Allows you to write it, but xarray still wants to use some weird cftime encoding (even though the original data did *not* have that), causing overflow when you try to reload. Do you know how to not use cftime on write, just use actual datetime values? i.e., this is not a kerchunk problem, its an xarray/zarr encoding thing.

**Comment by rsignell-usgs** [comment]
@martindurant so you don't think this has anything to do with the grib2 codec or reading grib2 files?

**Comment by keewis** [comment]
this completes successfully:
```python
subset = ds.assign_coords(time=ds.time.astype("datetime64[ns]")).isel(time=slice(-3,-1))
for var in subset.variables.values():
    var.encoding.pop("filters", None)
subset.to_zarr('foo.zarr', mode='w')

ds = xr.open_dataset("foo.zarr", engine="zarr").load()
```
`cftime` is only involved with dtypes that are not exactly `datetime64[ns]`, so casting `time` avoids that. `filters` raised errors when I tried to load the resulting `zarr`.

**Comment by rsignell-usgs** [comment]
@keewis , whoa, thank you!  I never would have figured this out!  Is this something that users should handle using this approach or should we handle it in xarray?

**Comment by martindurant** [comment]
Oh right, our filter to read GRIB. You would definitely need that to get any useful data out of the variables when reading from the original data, but indeed they make no sense at all for the newly created local zarr dataset. The GRIB_* attributes probably do carry useful information, though. I can't really tell whether popping them out as in the example affects the load or the save.

**Comment by keewis** [comment]
sorry, my bad, the attrs have nothing to do with this, and they are not affected by the `pop`, either.

**Comment by martindurant** [comment]
No, the filter is essential to read the GRIB files, but it makes no sense when writing out to zarr. xarray should have a way to write out without using the original filters; it's like saying "just because these data came from some special encoding, when I write them again, I want to use the same encoding". We do not, however, have a way to re-encode into GRIB-in-zarr, so xarray should allow us not to do this. Maybe there is such an option, the equivalent of pop for the filters, above?

**Comment by keewis** [comment]
I brought this up in the meeting yesterday. In short, removing or overriding the `filters` explicitly, or clearing `encoding` entirely are the only options we have at the moment. See also pydata/xarray#6323.

**Comment by rsignell-usgs** [comment]
@martindurant , in light of https://github.com/pydata/xarray/issues/6318#issuecomment-1059832700, do you think it would be best practice to not use datetime64 in our kerchunked zarr representation, even though it's allowed?

**Comment by martindurant** [comment]
>  Should we merge combine2?

It needs a lot of docs to explain how to use it - beyond the long docstring; and updating the existing examples to use it.

**Comment by rsignell-usgs** [comment]
okay, maybe I can get someone to work on that.
